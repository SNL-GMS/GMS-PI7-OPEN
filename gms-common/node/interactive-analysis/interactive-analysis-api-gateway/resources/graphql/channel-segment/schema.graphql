enum TimeSeriesType {
    WAVEFORM
    FK_SPECTRA
    DETECTION_FEATURE_MAP
}

# Represents a channel time segment comprised of one or more waveforms for a (derived or raw).
type ChannelSegment {

    # Unique identifier for the channel segment
    id: String!

    # Name of ChannelSegment
    name: String!

    # Type of timeseries
    timeseriesType: TimeSeriesType!

    # The type enumeration of the channel segment
    type: ChannelSegmentType

    # Unique identifier for the channel id
    channelId: String!

    # The start date/time (seconds since epoch) of the channel segment
    startTime: Float!

    # The end date/time ((seconds since epoch) of the channel segment
    endTime: Float!

    # The channel associated with the segment
    channel: ProcessingChannel

    # The collection of Waveforms associated with the channel segment
    timeseries: [Timeseries]

    # Provenance creation information for the channel segment
    creationInfo: CreationInfo!
}

# Input version of the time window parameters provided as input to the FK mutation API
input WindowParametersInput {

    # The type of window (e.g. Hanning)
    windowType: String!

    # The lead period in seconds (prior to the detection onset time)
    leadSeconds: Float!

    # The length of the window in seconds (including the lead and lag around the detection onset time)
    lengthSeconds: Float!

    #  The number of FK Spectra per second to fetch
    stepSize: Float!
}

input ContributingChannelsConfigurationInput {
    id: String!
    enabled: Boolean!
    name: String
}

# Holds the configuration used to calculate an Fk
  
input FkConfigurationInput {
    maximumSlowness: Float!
    mediumVelocity: Float!
    numberOfPoints: Float!
    normalizeWaveforms: Boolean!
    useChannelVerticalOffset: Boolean!
    contributingChannelsConfiguration: [ContributingChannelsConfigurationInput]!
    leadFkSpectrumSeconds: Float!
}
# Input parameters used to create a new FK data set
input FkInput {

    # The ID of the station the FK will be created for
    stationId: String!

    # Signal detection that Fk is created from
    signalDetectionId: String!

    # Signal detection hypothesis
    signalDetectionHypothesisId: String!

    # Phase of the signal detection
    phase: String!

    # The frequency band to compute the Fk on
    frequencyBand: FrequencyBandInput!

    # The window parameters to compute the Fk with
    windowParams: WindowParametersInput!

    # Configuration settings for the continous fk
    configuration: FkConfigurationInput
}

# Input to modify leadSeconds and return SD with leadFkSpectrum and PowerSpectrum populated
input FkLeadInput {
     # Signal detection that Fk is created from
    signalDetectionId: String!

    # Window Lead Seconds and Lead Fk Spectrum Select Seconds (movie)
    windowLeadSeconds: Float! # If lead is changed thru the widget
    leadFkSpectrumSeconds: Float!
}

# Input parameters that indicate which Fk (ChannelSegment TimeSeries)
# to set the reviewed flag
input MarkFksReviewedInput {
    # Signal detection that Fk is created from
    signalDetectionIds: [String]!

    # Boolean flag to set if reviewed by the Analyst
    reviewed: Boolean!
}

extend type Query {
    # Compute the Fk Thumbnails for UI Display. Should be called from UI after a
    # new Fk is called
    computeFkFrequencyThumbnails(fkInput: FkInput!): FkFrequencyThumbnailBySDId!
}

# Currently, no queries are provided. Fk data are accessed as members of
# AzslownessFeatureMeasurements on SignalDetectionHypotheses

# The following signal FK mutations are supported.
extend type Mutation {
    setFkWindowLead(leadInput: FkLeadInput!): AssociationChange

    # Compute the new FKs and set for each signal detection id
    computeFks(fkInput: [FkInput]!): AssociationChange

    # update reviewed flag 
    markFksReviewed(markFksReviewedInput: MarkFksReviewedInput!): AssociationChange

    # Calculate new Beam (replaces ArrivalTime FM Beam)
    computeBeam(fkInput: FkInput!): AssociationChange
}
