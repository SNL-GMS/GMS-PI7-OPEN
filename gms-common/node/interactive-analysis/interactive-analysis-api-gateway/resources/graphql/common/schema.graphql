# Common GraphQL schema definitions shared across the components of the
# API Gateway (e.g. waveform, workflow)

# Units used in DoubleValue part of feature prediction
enum Units {
    DEGREES
    SECONDS
    SECONDS_PER_DEGREE
    UNITLESS
}

# Represents Double fields for a double value representation.
type DoubleValue {
    value: Float # Optional Guidance is value can be a NaN i.e. FP predicted value
    standardDeviation: Float!
    units: Units!
}

# Represents Double fields for a double value representation input from UI.
input DoubleValueInput {
    value: Float!
    standardDeviation: Float!
    units: Units!
}

# Phase types
enum PhaseType {
    P
    S
    P3KPbc
    P4KPdf_B
    P7KPbc
    P7KPdf_D
    PKiKP
    PKKSab
    PKP2bc
    PKP3df_B
    PKSab
    PP_1
    pPKPbc
    PS
    Rg
    SKiKP
    SKKSac
    SKPdf
    SKSdf
    sPdiff
    SS
    sSKSdf
    Lg
    P3KPbc_B
    P5KPbc
    P7KPbc_B
    Pb
    PKKP
    PKKSbc
    PKP2df
    PKPab
    PKSbc
    PP_B
    pPKPdf
    PS_1
    SKKP
    SKKSac_B
    SKS
    SKSSKS
    sPKiKP
    SS_1
    SSS
    nNL
    P3KPdf
    P5KPbc_B
    P7KPbc_C
    PcP
    PKKPab
    PKKSdf
    PKP3
    PKPbc
    PKSdf
    pPdiff
    PPP
    pSdiff
    Sb
    SKKPab
    SKKSdf
    SKS2
    Sn
    sPKP
    SS_B
    SSS_B
    NP
    P3KPdf_B
    P5KPdf
    P7KPdf
    PcS
    PKKPbc
    PKP
    PKP3ab
    PKPdf
    Pn
    pPKiKP
    PPP_B
    pSKS
    ScP
    SKKPbc
    SKP
    SKS2ac
    SnSn
    sPKPab
    sSdiff
    NP_1
    P4KPbc
    P5KPdf_B
    P7KPdf_B
    Pdiff
    PKKPdf
    PKP2
    PKP3bc
    PKPPKP
    PnPn
    pPKP
    PPS
    pSKSac
    ScS
    SKKPdf
    SKPab
    SKS2df
    SP
    sPKPbc
    sSKS
    P4KPdf
    P5KPdf_C
    P7KPdf_C
    Pg
    PKKS
    PKP2ab
    PKP3df
    PKS
    PP
    pPKPab
    PPS_B
    pSKSdf
    Sdiff
    SKKS
    SKPbc
    SKSac
    SP_1
    sPKPdf
    sSKSac
    Sx
    tx
    N
    Px
    PKhKP
    LR
    pP
    UNKNOWN
}

# A date/time Type
scalar Date

# Represents a time range
# type TimeRange {
#     startTime: Float!
#     endTime: Float!
# }

# Input verion of a time range
input TimeRange {
    startTime: Float!
    endTime: Float!
}

# Represents a location specified using latitude (degrees), longitude (degrees),
# and altitude (kilometers).
type Location {

    # Latitude in degrees
    latDegrees: Float!
    
    # Longitude in degrees
    lonDegrees: Float!

    # Elevation in kilometers
    elevationKm: Float!
}

input LocationInput {

    # Latitude in degrees
    latDegrees: Float!
    
    # Longitude in degrees
    lonDegrees: Float!

    # Elevation in kilometers
    elevationKm: Float!
}

# Distance value's units degrees or kilometers
enum DistanceUnits {
    degrees
    km
}

# Enumerated list of source types used to compute distances to
enum DistanceSourceType {
    Event
    UserDefined
}

# Represents a distance measurement relative to a specified source location
type DistanceToSource {

    # The distance in kilometers or degrees to the source
    distance: Float!

    # The units distance was calculated
    distanceUnits: DistanceUnits

    # The source location
    sourceLocation: Location!

    # The type of the source the distance is measured to (e.g. and event)
    sourceType: DistanceSourceType!

    # Optional: the unique ID of the source object
    sourceId: String!

    # Which station distance to the source
    stationId: String!
}

input DistanceToSourceInput {
    # The source location
    sourceLocation: LocationInput

    # The type of the source the distance is measured to (e.g. and event)
    sourceType: DistanceSourceType

    # The Distance Units to use
    distanceUnits: DistanceUnits

    # Optional: the unique ID of the source object
    sourceId: String
}

# Represents a frequency range
type FrequencyBand {
    # Minimum frequency in Herz
    minFrequencyHz: Float!

    # Maximum frequency in Herz
    maxFrequencyHz: Float!
}

# Represents a frequency range input to a mutation
input FrequencyBandInput {
    # Minimum frequency in Herz
    minFrequencyHz: Float!

    # Maximum frequency in Herz
    maxFrequencyHz: Float!
}

# Enumeration representing the different types of stations in the monitoring network
enum StationType {

    Seismic3Component,
    Seismic1Component,
    SeismicArray,
    Hydroacoustic,
    HydroacousticArray,
    Infrasound,
    InfrasoundArray,
    Weather,
    UNKNOWN
}

# Type enumeration for channel segments
enum ChannelSegmentType {
    ACQUIRED
    RAW
    DETECTION_BEAM
    FK_BEAM
    FILTER
    FK_SPECTRA
}

# Provenance reference details for information added to the System
type InformationSource {

    # Unique ID for the source
    id: String!

    # Source organization
    organizationName: String!

    # The time associated with the information source
    referenceTime: Float!
}

# Enumeration of configuration status changes reflected in the configuration
enum ConfigurationInfoStatus {

    # Initial configuration
    Initial

    # The system's installed location for the item associated with the configuration was corrected
    CorrectedSystemLocation

    # The physical location of the ite associated with the configuration was changed
    MovedPhysicalLocation

    # The item associated with the configuration was decommissioned
    Decommissioned
}

# Provenance information associated with a configuration item in the system
type ConfigurationInfo {

    # The unique ID of the user who created the Configuration
    creatorId: String!

    # The time the configuration took effect in the real world (e.g. station came online)
    effectiveTime: Float!

    # The time the configuration was installed (and thus took effect) in the System
    installationTime: Float!

    # The status change reflected in the configuration values
    status: ConfigurationInfoStatus!

    # Comment associated with the configuration information
    comment: String
}

# Enumerated type of the actor (e.g. analyst, system) associated with a CreationInfo object
enum CreatorType {
    Analyst
    System
}

# Provenance information about the results of data processing completed by the System
# and actions completed by users of the system
type CreationInfo {
    # The unique ID of the creation info object
    id: String!

    # The time (seconds since Unix epoch) the associated processing result was created
    creationTime: Float!

    # The unique ID of the user who created the associated processing result
    creatorId: String!

    # The enumerated type of the creator (e.g. analyst, system)
    creatorType: CreatorType!

     # The name of the creation
    creatorName: String!

    # TODO
    # The processing step associated with the processing result
    # processingStep: ProcessingStep!

    # TODO
    # The software component associated with the processing result
    # softwareComponent: SoftwareComponent!
}

input CreationInfoInput {
    # Identifier for the creation info
    creatorName: String!
    
    # The time (seconds since Unix epoch) the associated processing result was created
    creationTime: Float!

    # The software information associated with the processing result
    softwareInfo: SoftwareInfoInput!
}

type SoftwareInfo {
    # Identifier for the creation info
    name: String!

    # The software version identifier
    version: String!
}

input SoftwareInfoInput {
    # Identifier for the creation info
    name: String!

    # The software version identifier
    version: String!
}

input ProcessingContextInput {
    analystActionReference: AnalystActionReferenceInput!
    processingStepReference: ProcessingStepReferenceInput!
    storageVisibility: String!
}

input AnalystActionReferenceInput {
    processingStageIntervalId: String!
    processingActivityIntervalId: String!
    analystId: String!
}

input ProcessingStepReferenceInput {
    processingStageIntervalId: String!
    processingSequenceIntervalId: String!
    processingStepId: String!
}

# Represents an alias for a station, channel, site, etc.
type Alias {

    # The alias name
    name: String!

    # Provenance information about the alias
    configurationInfo: ConfigurationInfo
}

# Represents calibration information associated with a waveform
type ProcessingCalibration {

    # Calibration factor
    factor: Float!

    # Calibration factor error
    factorError: Float!

    # Calibration period
    period: Float!

    # Calibration time shift
    timeShift: Float!
}

# Input version of calibration information associated with a waveform
input ProcessingCalibrationInput {

    # Calibration factor
    factor: Float!

    # Calibration factor error
    factorError: Float!

    # Calibration period
    period: Float!

    # Calibration time shift
    timeShift: Float!
}

# Represents a generic timeseries interface containing the fields common to the
# concrete timeseries data types defined in the system (e.g. Waveform, FkPowerSpectrum, DetectionFeatureMap)
interface Timeseries {
    # The start date/time (seconds since epoch) of the timeseries
    startTime: Float!

    # The sample rate (Hz) of the samples in the timeseries
    sampleRate: Float!

    # The sample count in the timeseries
    sampleCount: Float!
}

input TimeseriesInput {
    # The start date/time (seconds since epoch) of the timeseries
    startTime: String!
    
    # The sample rate (Hz) of the samples in the timeseries
    sampleRate: Float!
    
    # The sample count in the timeseries
    sampleCount: Float!
}

# Represents a time-bounded segment of waveform samples output from a channel (derived or raw).
# Waveforms implement the Timeseries interface
type Waveform implements Timeseries {
    # Unique identifier for the waveform
    # id: String!

    # The start date/time (seconds since epoch) of the waveform
    startTime: Float!

    # The sample rate (Hz) of the samples in the waveform
    sampleRate: Float!

    # The sample count in the waveform
    sampleCount: Float!

    # The time-ordered collection of waveform sample values
    values: [Float]
}

# Input version of a time-bounded segment of waveform samples output from a channel (derived or raw).
# Waveforms implement the Timeseries interface
input WaveformInput {
    
    # Unique identifier for the waveform
    id: String!

    # The start date/time (seconds since epoch) of the waveform
    startTime: Float!

    # The end date/time (seconds since epoch) of the waveform
    endTime: Float!    # The sample rate (Hz) of the samples in the waveform
    sampleRate: Float!

    # The sample count in the waveform
    sampleCount: Float!

    # The time-ordered collection of waveform sample values
    values: [Float]

    # Calibration information for the waveform
    calibration: ProcessingCalibrationInput
}

# The following queries are supported
type Query {
    # Dummy
    queryCount: Int
}

# The following mutations are supported
type Mutation {
    # Dummy
    mutationCount: Int
}

# The following subscriptions are supported
type Subscription {
    # Dummy
    subscriptionCount: Int
}

# Association Change object to return
type AssociationChange {
    events: [Event]
    sds: [SignalDetection]
}

 # Used when we need to return the original event mutated where performed on
 # in addition to the other data changed
type EventAndAssociationChange {
    event: Event!
    associationChange: AssociationChange!
}

#
enum LogLevel {
    INFO
    DEBUG
    WARNING
    ERROR
    DATA
}

type ClientLog {
    logLevel: LogLevel
    message: String!
    time: String
}

input ClientLogInput {
    logLevel: LogLevel!
    message: String!
    time: String
}

extend type Mutation {
    clientLog(clientLogInput: ClientLogInput): ClientLog
}