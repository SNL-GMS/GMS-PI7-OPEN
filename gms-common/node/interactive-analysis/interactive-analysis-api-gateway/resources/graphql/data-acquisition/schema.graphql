
# Represents the transfer gap fields used for gap status.
type FileGap {
    # Station Id
    stationName: String

    # Channel Id
    channelNames: [String]

    # Start time for transfer gap
    startTime: String

    # End time for transfer gap
    endTime: String

    # Duration for transfer gap
    duration: String

    # Location for transfer gap
    location: String

    # Priority for transfer gap
    priority: String
}

# The following queries are supported.
extend type Query {
    # Retrieve transfer file information by time range
    transferredFilesByTimeRange(timeRange: TimeRange!): [FileGap]
}

# Input to Reference Station mutations
input ReferenceStation {
    name: String!
    description: String!
    stationType: StationType!
    comment: String!
    source: InformationSourceInput!
    latitude: Float!
    longitude: Float!
    elevation: Float!
    actualChangeTime: String!
    systemChangeTime: String!
    aliases: [ReferenceAliasInput]!
}

input InformationSourceInput {
    originatingOrganization: String
    informationTime: String
    reference: String
}

input ReferenceAliasInput {
    id: String
    name: String
    status: StatusTypeName
    comment: String
    actualChangeTime: String
    systemChangeTime: String
}

enum StatusTypeName {
    INACTIVE,
    ACTIVE
}

type SaveStationResult{
    id: String!
    result: Boolean!
}

# The following Data Aquisition mutations are supported.
extend type Mutation {
    saveReferenceStation (input: ReferenceStation!): SaveStationResult
}