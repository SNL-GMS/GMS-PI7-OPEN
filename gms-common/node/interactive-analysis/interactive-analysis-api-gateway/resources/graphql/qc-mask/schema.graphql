# Represents a qc mask for a time segment for one or more channels.
type QcMask {

    # Unique identifier of QC mask
    id: String!

    # The id for the processing channel
    channelId: String!

    # QC Mask Version 
    qcMaskVersions: [QcMaskVersion]!

    # Current Mask Version
    currentVersion: QcMaskVersion!
},

# QC mask information
type QcMaskVersion {
    category: String!
    channelSegmentIds: [String]!
    creationInfo: CreationInfo!
    startTime: Float!
    endTime: Float!
    parentQcMasks: [QcMask]!
    rationale: String!
    type: String
    version: Int!
},

extend type Query {
    # Get all QcMasks for each channel that that fall within the start and endtime
    # If no channelIds are provided, Get all masks in the specified time range
    qcMasksByChannelId(timeRange: TimeRange!, channelIds: [String]): [QcMask]
},

# Input to QC Mask mutations
input QcMaskInput {
    timeRange: TimeRange

    category: String

    type: String

    rationale: String
}

# The following QC Mask mutations are supported.
extend type Mutation {
    # Creates new QC Masks
    createQcMask (channelIds: [String]!, input: QcMaskInput!): [QcMask]

    # Update an existing QC Mask using the provided input
    updateQcMask (qcMaskId: String!, input: QcMaskInput!): [QcMask]

    # Reject an existing QC Mask
    rejectQcMask (qcMaskId: String!, rationale: String): [QcMask]  
}

# The following subscription are supported.
extend type Subscription {
    # Subscribe for updates to QC masks on the specified channels in the
    # specified time range
    # If no parameters are provided, subscribe for all mask updates

    # Subscription for QC Masks created
    qcMasksCreated (timeRange: TimeRange, channelIds: [String]): [QcMask]   
}