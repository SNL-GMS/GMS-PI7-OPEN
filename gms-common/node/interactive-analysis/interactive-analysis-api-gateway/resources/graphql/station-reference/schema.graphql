# Represents a group of stations used for monitoring.
type Network {

    # Unique identifier for the network
    id: String!

    # Name of the network
    name: String

    # The monitoring organization the station is affiliated with
    monitoringOrganization: String

    # The geographic region associated with the network
    region: String

    # Provenance information about the network configuration
    configurationInfo: ConfigurationInfo

    # Source of network configuration information
    informationSource: InformationSource

    # Many-to-many association object between the network and associated stations
    # The association object is included to record provenance information
    networkMemberships: [NetworkMembership]
}

# Represents a many-to-may relationship between networks and stations
# with associated provenance information recorded in the associated
# configuration information
type NetworkMembership {

    # Unique identifier
    id: String!

    # Provenance information for this configuration item
    configurationInfo: ConfigurationInfo!

    # Station associated with the network
    station: Station!

    # Network associated with the station
    network: Network!
}

# Represents a many-to-may relationship between stations & sites
# with associated provenance information recorded in the associated
# configuration information
type StationMembership {

    # Unique identifier
    id: String!

    # Provenance information for this configuration item
    configurationInfo: ConfigurationInfo!

    # Site associated with the station
    site: Site!

    # Station associated with the site
    station: Station!
}

# An installation where monitoring sensors are installed.
# Multiple sensors can be installed at the same station.
type Station {

    # Unique identifier for the station
    id: String!

    # Name of the station
    name: String

    # Description of the station
    description: String

    # The type of station (seismic, hydroacoustic, infrasonic)
    stationType: StationType

    # Provenance information about the station configuration
    configurationInfo: ConfigurationInfo

    # Source of station configuration information
    informationSource: InformationSource

    # Location of the station
    location: Location!

    # Aliases for the site
    aliases: [Alias]

    # Many-to-many association object between the station and associated networks
    # The association object is included to record provenance information
    networkMemberships: [NetworkMembership]

    # Many-to-many association object between the station and associated sites
    # The association object is included to record provenance information
    stationMemberships: [StationMembership]

    # Latitude of the station
    latitude: Float

    # Longitude of the station
    longitude: Float

    # Elevation of the station
    longitude: Float
}



# Represents a many-to-may relationship between sites & channels
# with associated provenance information recorded in the associated
# configuration information
type SiteMembership {

    # Unique identifier
    id: String!

    # Provenance information for this configuration item
    configurationInfo: ConfigurationInfo!

    # Site associated with the channel
    site: Site!

    # channel associated with the site
    channel: Channel!
}


# Represents a physical installation (e.g., building, underground vault, borehole)
# containing a collection of Instruments that produce Raw Channel waveform data.
type Site {

    # Unique identifier for the site
    id: String!

    # The name of the site
    name: String

    # Provenance information about the site configuration
    configurationInfo: ConfigurationInfo

    # Source of site configuration information
    informationSource: InformationSource

    # Location of the site
    location: Location!

    # Aliases for the site
    aliases: [Alias]

    # Many-to-many association object between the site and associated stations
    # The association object is included to record provenance information
    stationMemberships: [StationMembership]

    # Many-to-many association object between the site and associated channels
    # The association object is included to record provenance information
    stationMemberships: [StationMembership]
}

# Represents an measurement instrument associated with a raw channel.
type Instrument {

    #  Unique identifier for the instrument
    id: String!

    # The name of the instrument manufacturer
    manufacturer: String

    # The name of the instrument model
    model: String

    # The instrument serial number
    serialNumber: String

    # Corner period of the instrument
    cornerPeriod: Float

    # Provenance information about the instrument configuration
    configurationInfo: ConfigurationInfo

    # Source of instrument configuration information
    informationSource: InformationSource
}

# Represents a source for unprocessed (raw) or processed (derived) time series data
# from a seismic, hydroacoustic, or infrasonic sensor
type Channel {

    # Unique identifier for the channel
    id: String!

    # Name of the channel
    name: String!

    # The type of the channel
    channelType: ChannelType!

    # Location code of the channel
    locationCode: String

    # The vertical angle (degrees) of the channel
    verticalAngle: Float

    # The horizontal angle (degrees) of the channel
    horizontalAngle: Float

    # The sample rate of the channel
    sampleRate: Float

    # The depth of the channel
    depth: Float

    # The instrument associated with the channel
    instrument: Instrument

    # Provenance information about the channel configuration
    configurationInfo: ConfigurationInfo

    # Source of channel configuration information
    informationSource: InformationSource

    # List of aliases for the channel
    aliases: [Alias]

}

# The following channel-related queries are supported.
extend type Query {

    # Retrieve the default channel list for use in automated and interactive analysis
    defaultChannels: [Channel]
}