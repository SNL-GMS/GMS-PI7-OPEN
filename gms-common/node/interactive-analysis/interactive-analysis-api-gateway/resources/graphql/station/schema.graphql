# Represents a group of stations used for monitoring.
# This is a simplified interface used for processing. A richer interface modeling
# station reference information is provided in the station reference API.
type ProcessingNetwork {

    # Unique identifier for the network
    id: String!

    # Name of the network
    name: String

    # The monitoring organization the network is affiliated with
    monitoringOrganization: String

    # The list of processing stations affiliated with the network
    stations: [ProcessingStation]
}

# Represents the station information fields used for station status.
type DataAcquisition {
  # Status for data acquisition
  dataAcquisition: String

  # Status for interactive processing
  interactiveProcessing: String

  # Status for automatic processing
  automaticProcessing: String
}

# Represents an installation of monitoring sensors.
# Multiple sensors can be installed at the same station.
# This is a simplified interface used for processing. A richer interface modeling
# station reference information is provided in the station reference API.
type ProcessingStation {

    # Unique identifier for the station
    id: String!

    # Name of the station
    name: String

    # The type of station (seismic, hydroacoustic, infrasonic)
    stationType: StationType

    # The reference location for the station
    location: Location

    # The description of the station
    description: String

    # The list of sites associated with the station
    sites: [ProcessingSite]

    # The default channel associated with the station
    defaultChannel: ProcessingChannel

    # The networks the station is associated with
    networks: [ProcessingNetwork]

    # The station information for Data Acquisition
    dataAcquisition: DataAcquisition

    # The latitude of the station
    latitude: Float

    # The longitude of the station
    longitude: Float

    # The elevation of the station
    elevation: Float

}

# Represents a physical installation (e.g., building, underground vault, borehole)
# containing a collection of Instruments that produce Raw Channel waveform data.
# This is a simplified interface used for processing. A richer interface modeling
# station reference information is provided in the station reference API.
type ProcessingSite {

    # Unique identifier for the site
    id: String!

    # The name of the site
    name: String

    # The location of the site
    location: Location!

    # The station the site is associated with
    station: ProcessingStation

    # The processing channels associated with the site
    channels: [ProcessingChannel]

    # The default channel associated with the site
    defaultChannel: ProcessingChannel
}

# Position information relative to a location
type Position {
  northDisplacementKm: Float
  eastDisplacementKm: Float
  verticalDisplacementKm: Float
}

# Represents a source for unprocessed (raw) or processed (derived) time series data
# from a seismic, hydroacoustic, or infrasonic sensor.
# This is a simplified interface used for processing. A richer interface modeling
# station reference information is provided in the station reference API.
type ProcessingChannel {

    # Unique identifier for the channel
    id: String!

    # Name of the channel
    name: String!

    # The type of the channel
    channelType: String!

    # Location code of the channel
    locationCode: String

    # The site associated with the channel
    site: ProcessingSite

    # The vertical angle (degrees) of the channel
    verticalAngle: Float

    # The horizontal angle (degrees) of the channel
    horizontalAngle: Float

    # Position information relative to the station's location
    position: Position

    # Actual change time
    actualTime: String

    # System change time
    systemTime: String

    # The sample rate of the channel (Hertz) of the channel
    sampleRate: Float

    # The depth of the channel
    depth: Float

    # QC Masks for the channel
    qcMasks(timeRange: TimeRange): [QcMask]
}

# The following channel-related queries are supported.
extend type Query {

    # Retrieve the stations in a specified network
    stationsByNetwork(networkName: String!): [ProcessingStation]

    # Retrieve the default station list for use in automated and interactive analysis
    defaultStations: [ProcessingStation]

    # Retrieve a list of processing channels by ID
    channelsById(ids: [String]): [ProcessingChannel]

    # Distance to a source (e.g. event, user-specified location)
    distanceToSourceForDefaultStations(distanceToSourceInput: DistanceToSourceInput): [DistanceToSource]
}
