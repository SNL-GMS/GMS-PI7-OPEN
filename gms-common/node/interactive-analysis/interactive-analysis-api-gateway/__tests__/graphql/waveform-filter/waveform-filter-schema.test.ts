import { delayExecution } from '../../../src/ts/util/delay-execution';
import { objectToGraphQLString } from '../../../src/ts/util/object-to-graphql';
import { graphql } from 'graphql';
import { schema } from '../../../src/ts/schema';
import { configProcessor } from '../../../src/ts/config/config-processor';
import { WaveformFilter } from '../../../src/ts/waveform-filter-definition/model';

const delayMs = 500;

describe('waveform-filter query', () => {
    let filterData: WaveformFilter[];
    test('should have a waveform filter query and mutation', async () => {
        const numberOfDefaultFilters = 6;
        const newSampleRate = 555;
        const waveformFilterIds: string[] = configProcessor.getConfigByKey('waveformFilterIds');

        expect(waveformFilterIds).toBeDefined();
        expect(waveformFilterIds).toHaveLength(numberOfDefaultFilters);

        const query = `
        query waveformFilterQuery {
            defaultWaveformFilters{
                id
                name
                description
                filterType
                filterPassBandType
                lowFrequencyHz
                highFrequencyHz
                order
                filterSource
                filterCausality
                zeroPhase
                sampleRate
                sampleRateTolerance
                groupDelaySecs
                validForSampleRate
            }
        }
        `;

        const rootValue = {};
        const result = await delayExecution(() =>  graphql(schema, query, rootValue), delayMs);
        // const { data } = result;
        filterData = result.data.defaultWaveformFilters;

        // expect(waveformFilterIds.length).toEqual(data.length);
        expect(waveformFilterIds.length).toEqual(filterData.length);
        // Compare response to snapshot
        expect(filterData).toMatchSnapshot();

        filterData[0].sampleRate = newSampleRate;

        const mutation = `
        mutation updateWfFilters {
            updateWfFilter (input: {${objectToGraphQLString(filterData[0], '')}}) {
                id
                name
                description
                filterType
                filterPassBandType
                lowFrequencyHz
                highFrequencyHz
                order
                filterSource
                filterCausality
                zeroPhase
                sampleRate
                sampleRateTolerance
                groupDelaySecs
                validForSampleRate
            }
          }
        `;

        const mutationRootValue = {};
        const mutationResult = await delayExecution(() =>  graphql(schema, mutation, mutationRootValue), delayMs);

        // Compare mutation return to the expected change made
        expect(mutationResult.data.updateWfFilter[0].sampleRate).toEqual(newSampleRate);
    });
});
