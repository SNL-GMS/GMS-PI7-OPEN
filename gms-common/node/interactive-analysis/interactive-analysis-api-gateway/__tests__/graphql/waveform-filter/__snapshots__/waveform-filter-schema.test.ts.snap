// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`waveform-filter query should have a waveform filter query and mutation 1`] = `
Array [
  Object {
    "description": "Hamming FIR Filter Band Pass, 0.70-2.00 Hz",
    "filterCausality": "CAUSAL",
    "filterPassBandType": "BAND_PASS",
    "filterSource": "SYSTEM",
    "filterType": "FIR_HAMMING",
    "groupDelaySecs": 1.2,
    "highFrequencyHz": 2,
    "id": "48fd578e-e428-43ff-9f9e-62598e7e6ce6",
    "lowFrequencyHz": 0.7,
    "name": "HAM FIR BP 0.70-2.00 Hz",
    "order": 48,
    "sampleRate": 20,
    "sampleRateTolerance": 0.05,
    "validForSampleRate": "true",
    "zeroPhase": "false",
  },
  Object {
    "description": "Hamming FIR Filter Band Pass, 0.70-2.00 Hz",
    "filterCausality": "CAUSAL",
    "filterPassBandType": "BAND_PASS",
    "filterSource": "SYSTEM",
    "filterType": "FIR_HAMMING",
    "groupDelaySecs": 0.6,
    "highFrequencyHz": 2,
    "id": "0351d87a-fde9-43e2-8754-84a1d797fbc1",
    "lowFrequencyHz": 0.7,
    "name": "HAM FIR BP 0.70-2.00 Hz",
    "order": 48,
    "sampleRate": 40,
    "sampleRateTolerance": 0.05,
    "validForSampleRate": "true",
    "zeroPhase": "false",
  },
  Object {
    "description": "Hamming FIR Filter Band Pass, 1.00-3.00 Hz",
    "filterCausality": "CAUSAL",
    "filterPassBandType": "BAND_PASS",
    "filterSource": "SYSTEM",
    "filterType": "FIR_HAMMING",
    "groupDelaySecs": 1.2,
    "highFrequencyHz": 3,
    "id": "56893621-d9a1-4cfd-830a-13969f9d2aad",
    "lowFrequencyHz": 1,
    "name": "HAM FIR BP 1.00-3.00 Hz",
    "order": 48,
    "sampleRate": 20,
    "sampleRateTolerance": 0.05,
    "validForSampleRate": "true",
    "zeroPhase": "false",
  },
  Object {
    "description": "Hamming FIR Filter Band Pass, 1.00-3.00 Hz",
    "filterCausality": "CAUSAL",
    "filterPassBandType": "BAND_PASS",
    "filterSource": "SYSTEM",
    "filterType": "FIR_HAMMING",
    "groupDelaySecs": 0.6,
    "highFrequencyHz": 3,
    "id": "8bab76bd-db1a-4226-a151-786577adb3d8",
    "lowFrequencyHz": 1,
    "name": "HAM FIR BP 1.00-3.00 Hz",
    "order": 48,
    "sampleRate": 40,
    "sampleRateTolerance": 0.05,
    "validForSampleRate": "true",
    "zeroPhase": "false",
  },
  Object {
    "description": "Hamming FIR Filter Band Pass, 4.00-8.00 Hz",
    "filterCausality": "CAUSAL",
    "filterPassBandType": "BAND_PASS",
    "filterSource": "SYSTEM",
    "filterType": "FIR_HAMMING",
    "groupDelaySecs": 1.19999999999,
    "highFrequencyHz": 8,
    "id": "db5e61a0-b8dc-48a7-b56c-5e497925e89c",
    "lowFrequencyHz": 4,
    "name": "HAM FIR BP 4.00-8.00 Hz",
    "order": 48,
    "sampleRate": 20,
    "sampleRateTolerance": 0.05,
    "validForSampleRate": "true",
    "zeroPhase": "false",
  },
  Object {
    "description": "Hamming FIR Filter Band Pass, 4.00-8.00 Hz",
    "filterCausality": "CAUSAL",
    "filterPassBandType": "BAND_PASS",
    "filterSource": "SYSTEM",
    "filterType": "FIR_HAMMING",
    "groupDelaySecs": 0.6,
    "highFrequencyHz": 8,
    "id": "b92d6ade-3e69-42b6-82e5-f9290f50120f",
    "lowFrequencyHz": 4,
    "name": "HAM FIR BP 4.00-8.00 Hz",
    "order": 48,
    "sampleRate": 40,
    "sampleRateTolerance": 0.05,
    "validForSampleRate": "true",
    "zeroPhase": "false",
  },
]
`;
